### PR TITLE
Traverse commit phase effects iteratively

### PR DESCRIPTION
**Pull request by acdlite**
Initial implementation done. It's in a working state, but the diff is large so I want to do a tidying pass and go through it again line-by-line. Maybe remove some duplication.

Would appreciate a thorough review. It's a lot of copy-paste-tweak, so it's likely I missed something.

## Description

We suspect that using the JS stack to traverse through the tree in the commit phase is slower than traversing iteratively. This changes the commit phase to use while loops and the fiber `return` pointer, instead of recursion.

There are a few places in the commit phase where we'd like to read data from the stack, e.g. to check if we're inside a hidden tree. In the render phase, we maintain a virtual "stack" on the heap. We could do that in the commit phase, too, but since the commit phase is synchronous, and most nodes do not need access to the stack, I've chosen to continue using the JS stack (recursion) whenever we hit a node that "provides" data to children. This is pretty rare, though, so it shouldn't have too much of an impact; we were already using this hybrid recursive-iterative strategy in `hideOrUnhideAllChildren`.

I've kept the recursive implementation behind a flag, both so we have the option to run an experiment comparing the two, and so we can revert it easily later if needed.

## To Do

- [x] Move rest of commit phase logic ReactFiberCommitWork
- [x] Snapshot phase
- [x] Mutation phase
- [x] Layout phase
- [x] Passive phase
- [x] Tidy up

### CODE REVIEW

**Comment by bvaughn** [comment]
Looks like another build variant the reconciler benchmark might want to compare in the short term, @rickhanlonii.

**Review by rickhanlonii — COMMENTED** [review]
The algos and overall strategy look good to me 👍

Will leave the detailed review to @bvaughn, who has context.

**Comment by acdlite** [comment]
@bvaughn Ok this is ready for you to look at

File: packages/react-reconciler/src/ReactFiberCommitWork.new.js [hunk 1]
```diff
@@ -660,18 +1115,229 @@ function recursivelyCommitLayoutEffects(
...
+        enableProfilerCommitHooks &&
+        finishedWork.tag === Profiler
+      ) {
+        const prevProfilerOnStack = nearestProfilerOnStack;
+        nearestProfilerOnStack = finishedWork;
+
+        let child = firstChild;
+        while (child !== null) {
+          nextEffect = child;
+          iterativelyCommitLayoutEffects_begin(child, finishedRoot);
```
Comment (bvaughn) [inline packages/react-reconciler/src/ReactFiberCommitWork.new.js:1157]: This seems like a reasonable approach 👍

File: packages/react-reconciler/src/ReactFiberCommitWork.new.js [hunk 2]
```diff
@@ -660,18 +1115,229 @@ function recursivelyCommitLayoutEffects(
...
+          child = child.sibling;
+        }
+        nextEffect = finishedWork;
+
+        if ((finishedWork.flags & LayoutMask) !== NoFlags) {
```
Comment (bvaughn) [inline packages/react-reconciler/src/ReactFiberCommitWork.new.js:1162]: Shouldn't have any practical implications, but we check `LayoutMask` in the iterative branch and `Update | Callback` in the recursive branch. Might be worth updating the latter to also check `LayoutMask`.

**Review by bvaughn — APPROVED** [review]
This change looks good. The new iterative methods that have to account for Profiler (and soon Offscreen) are a lot less readable in the iterative forks though :(

We aren't actually running any tests again the old recursive methods anymore, are we? That seems potentially problematic.

**Comment by acdlite** [comment]
> We aren't actually running any tests again the old recursive methods anymore, are we? That seems potentially problematic.

That's right, since the new reconciler is only enabled in the www variant build. I did turn it on locally to manually test, which can continue to do until the new reconciler is merged into the old. Then we can use `__VARIANT__`.

File: packages/shared/forks/ReactFeatureFlags.www.js [hunk 3]
```diff
@@ -84,6 +84,8 @@ export const enableDiscreteEventFlushingChange = true;
 // to the correct value.
 export const enableNewReconciler = __VARIANT__;
 
+export const enableRecursiveCommitTraversal = false;
```
Comment (rickhanlonii) [inline packages/shared/forks/ReactFeatureFlags.www.js:87]: Should this be either true or in the dynamic values?

Comment (bvaughn) [reply packages/shared/forks/ReactFeatureFlags.www.js:87]: See this comment https://github.com/facebook/react/pull/20094#issuecomment-717469662

Andrew's plan was to split iterative and recursive between old and new forks (once we merge the new fork into the old). For now, we're turning on iterative only for the new build (to see if that fixes the top line metrics).

Does make me a little uneasy that we don't have automated tests covering the new + recursive combination, but if we're able to merge new -> old soon, hopefully it's ok?

Comment (rickhanlonii) [reply packages/shared/forks/ReactFeatureFlags.www.js:87]: Yeah ignore me, I'm a clown. This forces iterative mode which is what we want.

**Comment by maraisr** [comment]
In layman terms, what does this PR mean? "We haz de fasta appz"?

**Comment by bvaughn** [comment]
PR = pull request

**Comment by rickhanlonii** [comment]
@maraisr this PR reworks the algorithms that fire effects and mutations in the commit phase to test a theory for why perf regressed in one of our algorithms that hasn't been released yet. These changes don't have any user-facing impact, and is just part of our research for now.

**Comment by maraisr** [comment]
Still mega interesting though!
